### PR TITLE
[9.0][FIX] fix overtime computation

### PR DESCRIPTION
**Pull request by robinkeunen**
**Before:** 
1. Wrong computation of overtime if an employee does not work a day he was supposed to work on.
ex: Employee L usually works from Monday to Thursday. This week, L decides to work on Friday instead of Wednesday. In the following example, the system computes overtime for Friday without computing negative overtime for Wednesday.

Screenshot 2020-01-22 at 16 01 14

2. total_overtime is computed on each timesheet, these can differ as time goes on.

**After**

1. the module takes all days over sheet range into account regardless if period exist for these days.
2. total_overtime is now computed on the employee.

### CODE REVIEW

File: hr_timesheet_overtime/models/hr_timesheet_sheet.py [hunk 1]
```diff
@@ -30,7 +30,7 @@ class HrTimesheetSheet(models.Model):
     total_overtime = fields.Float(
         "Overtime Total",
         readonly=True,
-        compute="_compute_total_overtime",
+        related="employee_id.total_overtime",
```
Comment (vvrossem) [inline hr_timesheet_overtime/models/hr_timesheet_sheet.py:33]: TIL how to use `related` :+1:

File: hr_timesheet_overtime/models/hr_employee.py [hunk 2]
```diff
@@ -28,3 +29,18 @@ class HrEmployee(models.Model):
...
+        """
+        for employee in self:
+            sheets = self.env["hr_timesheet_sheet.sheet"].search(
+                [
+                    ("employee_id", "=", employee.id),
+                    ("date_from", ">", employee.overtime_start_date),
+                ]
+            )
+            overtime = sum(sheet.timesheet_overtime for sheet in sheets)
+            employee.total_overtime = employee.initial_overtime + overtime
```
Comment (vvrossem) [inline hr_timesheet_overtime/models/hr_employee.py:46]: It makes indeed more sense to store `total_overtime` on the employee

File: hr_timesheet_overtime/models/hr_timesheet_sheet.py [hunk 3]
```diff
@@ -48,38 +49,14 @@ def _compute_timesheet_overtime(self):
         for sheet in self:
             ts_overtime = 0.0
             total_timesheet_period = sheet.get_total_timesheet_period()
-            for date in total_timesheet_period:
-                daily_wh = sheet.get_working_hours(date=date)
-                daily_overtime = total_timesheet_period[date] - daily_wh
-                ts_overtime += daily_overtime
+            for date, total_timesheet in total_timesheet_period.items():
```
Comment (vvrossem) [inline hr_timesheet_overtime/models/hr_timesheet_sheet.py:52]: better indeed to use `items()` to iterate over tuples

File: hr_timesheet_overtime/models/hr_timesheet_sheet.py [hunk 4]
```diff
@@ -124,7 +101,16 @@ def get_total_timesheet_period(self):
...
         )
-        return {ts.name: ts.total_timesheet for ts in ts_day}
+        date_to = fields.Date.from_string(self.date_to)
+        date_from = fields.Date.from_string(self.date_from)
+
+        nb_days = (date_to - date_from).days + 1
+        periods = {
+            fields.Date.to_string(date_from + timedelta(days=d)): 0.0
+            for d in range(0, nb_days)
+        }
```
Comment (vvrossem) [inline hr_timesheet_overtime/models/hr_timesheet_sheet.py:111]: Nicely done !

**Review by vvrossem — APPROVED** [review]
Thank you for improving this module. I understand better now how to use `related` and learnt a new usage of `timedelta`

**Comment by robinkeunen** [comment]
@vvrossem Thank you for the review. I suggest merging all validated PR then making a new one to get Travis back to green. I can help with that.

Manuel is taking the lead on Coopaname's development so I can be available to help on provelo.

**Pull request merged**